excel/src/main/java/org/apache/metamodel/excel/ExcelUtils.java

```diff
@@ -232,53 +249,119 @@ public static String getCellValue(Workbook wb, Cell cell) {
+        }
+        return errorResult;
+    }
+
+    private static Object getCellValueChecked(final Workbook workbook, final Cell cell,
+            final ColumnType expectedColumnType) {
+        final Object value = getCellValueAsObject(workbook, cell);
+        if (value == null || value.getClass().equals(expectedColumnType.getJavaEquivalentClass())) {
+            return value;
+        }
+
+        // Don't log when an Integer value is in a Double column type
+        if (!(value.getClass().equals(Integer.class) && expectedColumnType
+                .getJavaEquivalentClass()
+                .equals(Double.class)) && logger.isWarnEnabled()) {
+            logger
+                    .warn("Cell ({},{}) has the value '{}' of data type '{}', which doesn't match the detected "
+                            + "column's data type '{}'. This cell gets value NULL in the DataSet.", cell.getRowIndex(),
+                            cell.getColumnIndex(), value, value.getClass().getSimpleName(), expectedColumnType);
+        }
+        return null;
+    }
+
+    private static String getFormulaCellValue(Workbook workbook, Cell cell) {
         // first try with a cached/precalculated value
         try {
             double numericCellValue = cell.getNumericCellValue();
-            // TODO: Consider not formatting it, but simple using
-            // Double.toString(...)
-            return _numberFormat.format(numericCellValue);
+            return getNumericCellValueAsString(cell.getCellStyle(), numericCellValue);
         } catch (Exception e) {
             if (logger.isInfoEnabled()) {
                 logger.info("Failed to fetch cached/precalculated formula value of cell: " + cell, e);
             }
         }
 
         // evaluate cell first, if possible
+        final Cell evaluatedCell = getEvaluatedCell(workbook, cell);
+        if (evaluatedCell != null) {
+            return getCellValue(workbook, evaluatedCell);
+        } else {
+            // last resort: return the string formula
+            return cell.getCellFormula();
+        }
+    }
+
+    private static Object getFormulaCellValueAsObject(final Workbook workbook, final Cell cell) {
+        // first try with a cached/precalculated value
+        try {
+            return getDoubleAsNumber(cell.getNumericCellValue());
+        } catch (final Exception e) {
+            if (logger.isInfoEnabled()) {
+                logger.info("Failed to fetch cached/precalculated formula value of cell: " + cell, e);
+            }
+        }
+
+        // evaluate cell first, if possible
+        final Cell evaluatedCell = getEvaluatedCell(workbook, cell);
+        if (evaluatedCell != null) {
+            return getCellValueAsObject(workbook, evaluatedCell);
+        } else {
+            // last resort: return the string formula
+            return cell.getCellFormula();
+        }
+    }
+
+    private static Cell getEvaluatedCell(final Workbook workbook, final Cell cell) {
```

Comment:
Consider renaming this to `evaluateCell`.

Reply:
sure